packages/pg-transaction/src/index.ts

```diff
@@ -0,0 +1,38 @@
+import type { Client, Pool, PoolClient } from 'pg'
+
+function isPoolClient(clientOrPool: Client | PoolClient): clientOrPool is PoolClient {
+  return 'release' in clientOrPool
+}
+
+function isPool(clientOrPool: Client | Pool): clientOrPool is Pool {
+  return 'idleCount' in clientOrPool
+}
+
+async function transaction<T>(
+  clientOrPool: Client | Pool,
+  cb: (client: Client | PoolClient) => Promise<T>
+): Promise<T> {
+  let client: Client | PoolClient
+  if (isPool(clientOrPool)) {
+    // It's a Pool
+    client = await clientOrPool.connect()
+  } else {
+    // It's a Client
+    client = clientOrPool as Client
+  }
+  await client.query('BEGIN')
+  try {
+    const result = await cb(client)
+    await client.query('COMMIT')
+    return result
+  } catch (error) {
+    await client.query('ROLLBACK')
+    throw error
+  } finally {
+    if (isPoolClient(client)) {
+      client.release()
```

Comment:
If the rollback failed, `client` could be returned to the pool while still in a transaction or in a broken-connection state. I don’t know if the former is actually possible in practice (`ROLLBACK` failing without breaking the connection), but the latter definitely is (when `error` doesn’t crash the process, e.g. because the user is attaching `error` listeners to all clients using the pool’s `connect` event). Might want to add another `catch` for the `ROLLBACK` and `.release(rollbackError)`.

Reply:
oh that's a good point. I'll take stab at this...I think I can repro by `pg_terminate_backend` the active in transaction client backed from within the transaction itself. That should disconnect it and make it fail to be able to send `rollback` - anyways....I'll test a few more failure modes for both this comment and the other one about the error listener dance and see if I can make it throw an uncaught anywhere else.